### DISCUSSION TITLE
Question on release specific improvements

### DISCUSSION THREAD

**Discussion by nvysyaraju-sift** [nielsbasjes/yauaa#1516] [Q&A] [answered]
Hello @nielsbasjes, just have a couple of questions on the improvements here.

- V 7.21.0
    - New/improved detections:
        - Report Samsung DEX (Desktop Experience) as Tablet
- V 7.26.0
    - New/improved detections:
        - Report the agent version 'NULL' as '??'
       
As part of 7.21.0, one of the detection improvements say that the UA string with Samsung DEX as device are considered under `Tablet` DeviceClass but I do see it being detected as Tablet and Phone in older UA versions such as` 7.19.1` (I tried parsing it with this older version using the docker based webservlet). Just for clarity, does it detect `Samsung DEX` as any other DeviceClass type before? Also does this improvement will detect it only as `Tablet` as opposed to other possible classes? 

As part of 7.26.0, one of the improvments is reporting `NULL` agent version as `??`, unknown versions are always represented as `??` right? When do we see NULL version cases from the parser before this version update?

**Comment by nielsbasjes** [comment, marked as answer]
The Samsung DEX [change](https://github.com/nielsbasjes/yauaa/commit/999746001c8f7c7a93400c3523fbb2abb3a40973) is one of those really strange edgecases. When I actually got an S22 from work I was able to record and test what really happens when you activate DEX.

- Before this change it would have always been reported as a Phone ... because technically it IS a phone.
- After this change; when DEX is active (i.e. you have connected the phone to a desktop screen and the whole thing is behaving as a tablet) it is now reported as a Tablet instead because you have physically modified the device and you are now using it with a big screen.

The NULL version [change](https://github.com/nielsbasjes/yauaa/commit/07a2ff93cd62c6c8548ab72acd66e85485c158e1):
- There are actually useragents like this `Mozilla/5.0 (Windows NT 10.0; rv:68.0) Gecko/20100101 Firefox/NULL` which put `NULL` as the version and create this effect.

**Reply by nvysyaraju-sift** [reply]
Thank you for the context on these points.

So for the NULL version change, it will report it as `??`, which is same as the unknown version value we see for all the UA string for which browser/os version is not known right?

**Reply by nvysyaraju-sift** [reply]
Also can you give context on what does `Samsung Browser 24 now does Frozen UA` mean?

7.26.1
    - New/improved detections:
        - Samsung Browser 24 now does Frozen UA

**Reply by nielsbasjes** [reply]
The Samsung browser changed the behavior since that version so thar needed to be handled.

And yes, the null is reported as ?? because it is a different way of not knowing the version.

**Reply by nvysyaraju-sift** [reply]
> The Samsung browser changed the behavior since that version so thar needed to be handled.

Ohh okay but what did the improvement mean? Particularly what does `Samsung Browser 24 now does Frozen UA` mean in that improvment?

**Reply by nielsbasjes** [reply]
The Chromium based browsers have been changing the UserAgent to reduce the information encoded in it. For example the OS version has been frozen to a fixed value.  With this version this browser started to use the same "frozen" pattern.

**Reply by nielsbasjes** [reply]
See this https://yauaa.basjes.nl/expect/manipulations/#reducingfreezing-the-useragent
